ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -57,141 +53,143 @@ public CloudStorageCompactor(CloudDestination cloudDestination, CloudConfig clou
+      3. Shutdown thread: It initiates a graceful shutdown of worker and main threads.
+     */
+
+    // Shutdown worker threads, though this may not have an effect if workers are blocked on network calls
     cloudDestination.stopCompaction();
-    // VcrServer shuts down us before the destination.
-    boolean success = false;
-    try {
-      success = doneLatch.get().await(shutDownTimeoutSecs, TimeUnit.SECONDS);
-    } catch (InterruptedException ex) {
-    }
-    executorService.shutdown();
-    if (success) {
-      logger.info("Compactor shut down successfully.");
-      shutDown.set(true);
-    } else {
-      logger.warn("Timed out or interrupted waiting for operations to finish.  If cloud provider uses separate stores"
-          + " for data and metadata, some inconsistencies may be present.");
-      vcrMetrics.compactionShutdownTimeoutCount.inc();
-    }
+
+    /*
+      Shutdown executor.
+      The arbitrary wait period is proportional to the number of worker threads, instead of a fixed timeout.
+      It returns if all workers end before the timeout.
+      This wait is merely an attempt to allow the worker threads to gracefully exit. We will force a shutdown later.
+      All workers are daemons and JVM _will_ exit when only daemons remain.
+      Any data inconsistencies must be resolved separately, but not by trying to predict the right shutdown timeout.
+      cloudBlobCompactionShutdownTimeoutSecs is useful for reducing test shutdown times.
+    */
+    Utils.shutDownExecutorService(executorService, cloudConfig.cloudBlobCompactionShutdownTimeoutSecs,
+        TimeUnit.SECONDS);
   }
 
   /**
    * @return whether the compactor is shutting down.
    */
   boolean isShutDown() {
-    return shutDown.get();
+    return this.executorService.isShutdown();
   }
 
-  /**
-   * Purge the inactive blobs in all managed partitions.
-   * @return the total number of blobs purged.
-   */
-  public int compactPartitions() {
-    if (partitions.isEmpty()) {
-      logger.info("Skipping compaction as no partitions are assigned.");
+  protected boolean isPartitionOwned(PartitionId partitionId) {
+    return partitions.contains(partitionId);
+  }
+
+  protected class CompactionTask implements Callable<Integer> {
+    private final PartitionId partitionId;
+    CompactionTask(PartitionId partitionId) {
+      /*
+        Save a reference to partitionId since it's a local variable in the submitting loop
+        and will go out of scope before the task begins.
+       */
+      this.partitionId = partitionId;
+    }
+    @Override
+    public Integer call() throws Exception {
+      /*
+        Jobs wait for execution until workers become available.
+        If 100 jobs are submitted to 5 workers, only 5 jobs run while the rest wait.
+        Partitions can be reassigned while jobs are in the queue.
+        Verify ownership of the partition before job initiation.
+       */
+      if (!isPartitionOwned(partitionId)) {
```

Comment:
This the callable submitted to the executor. 
If shutdown has happened, no new callable can be submitted. A shutdown check doesn't help this case because the callable will never get to it.
If shutdown has happened, waiting callable will be cancelled. A shutdown check doesn't help this case because the callable will never get to it.
If shutdown has happened, running callable will be interrupted and terminated. A shutdown check doesn't help this case because the callable will be interrupted either before it runs cloudDestination.compactPartition() or while it is running cloudDestination.compactPartition() or will gracefully exit cloudDestination.compactPartition().

Reply:
ExecutorService has shutdown and shutdownNow.
In our util.shutDownExecutorService, it calls shutdown first and wait. After timeout, then it calls shutdownNow.

When ExecutorService.shutdown happened,
1. if new callable submitted, it'll be rejected by Executor.
2. for the running callable, it continues the execution.
3. for the waiting callable, they will still be executed since it's in the queue already. 
In that case, the 1000 callable will all be executed.

When ExecutorService.shutdownNow happened, then the running callable will be interrupted.

Reply:
I got a better idea, lemme code it up